concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -1138,79 +1140,61 @@ handleDeployCredential ::
-              Just arsInfos -> do
-                cryptoParams <- lift getCryptoParams
-                -- we have two options. One is that we are deploying a credential on an existing account.
-                case ID.cdvPublicKeys ncdv of
-                  ID.CredentialPublicKeys keys _ -> do
-                    -- account does not yet exist, so create it, but we need to be careful
-                    when (null keys || length keys > 255) $ throwError $ Just AccountCredentialInvalid
-                    let aaddr = ID.addressFromRegId regId
-                    case ID.values cdi of
-                      Nothing -> throwError $ Just AccountCredentialInvalid
-                      Just cdv -> do
-                        -- this check is extremely unlikely to fail (it would amount to a hash collision since
-                        -- we checked regIdEx above already).
-                        accExistsAlready <- lift (addressWouldClash aaddr)
-                        let check = AH.verifyCredential cryptoParams ipInfo arsInfos cdiBytes (Left messageExpiry)
-                        unless (not accExistsAlready && check) $ throwError $ Just AccountCredentialInvalid
-                        -- Add the account to the state, but only if the credential was valid and the account does not exist
-                        _ <- lift (createAccount cryptoParams aaddr cdv)
-                        mkSummary (TxSuccess [AccountCreated aaddr, CredentialDeployed{ecdRegId=regId,ecdAccount=aaddr}])
-  case res of
-    Left err -> return (TxInvalid <$> err)
-    Right ts -> return (Just ts)
-
+handleDeployCredential accCreation@AccountCreation{messageExpiry=messageExpiry, credential=cdi} cdiHash = do
+    res <- runExceptT $ do
+      liftedCm <- lift getChainMetadata
+      let ts = slotTime liftedCm
+      -- check that the transaction is not expired
+      when (transactionExpired messageExpiry ts) $ throwError (Just ExpiredTransaction)
+      remainingEnergy <- lift getRemainingEnergy
+      let cost = Cost.deployCredential (ID.credentialType cdi) (ID.credNumKeys . ID.credPubKeys $ cdi)
+      when (remainingEnergy < cost) $ throwError Nothing
+      let mkSummary tsResult = do
+            tsIndex <- lift bumpTransactionIndex
+            return $ TxValid $ TransactionSummary{
+              tsSender = Nothing,
+              tsHash = cdiHash,
+              tsCost = 0,
+              tsEnergyCost = cost,
+              tsType = TSTCredentialDeploymentTransaction (ID.credentialType cdi),
+              ..
+              }
+
+      let regId = ID.credId cdi
+      let aaddr = ID.addressFromRegId regId
+      -- We always need to make sure that the account was not created in between
+      -- the transaction was received and the actual execution.
+      accExistsAlready <- lift (addressWouldClash aaddr)
+      when accExistsAlready $ throwError $ Just $ DuplicateAccountRegistrationID regId
```

Comment:
This is not the error we currently return and is technically not the correct one.

Reply:
I think we need to separately check that the credential registration does not exist already since it might exist at this point.

Even assuming no hash collisions on the 29 bytes it is possible that a credId exists, but there is no clashing account.

Reply:
Introduced a check for the registration id and also fixed the error for clashing account addresses.